### PR TITLE
Add NYM Swap (nymswap) reporting plugin

### PR DESCRIPTION
**Pull request by j0ntz**
### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [x] Yes
- [ ] No

### Dependencies

none

### Description

Add a new reports-server partner plugin `nymswap` so NYM swaps surface in the reports
pipeline. Asana: https://app.asana.com/0/1215088146871429/1216728226940270

The reporting contract is confirmed against NYM's live "Edge Partner" API (OpenAPI at
https://nym-swap-api.nymtech.cc/api/docs/) and a live query run under the GUI's existing
NYM swap key (`NYM_SWAP_INIT.apiKey`), which had real order history to read.

What it does:
- `src/partners/nym.ts` - `queryNym` walks `GET /api/partner/v1/reports/transactions`
  (auth `x-api-key`, params `startDate`/`endDate`/`limit<=500`/`cursor`, envelope
  `{ transactions, nextCursor }`), following `nextCursor` until null with a 5-day lookback,
  retry/snooze on throttling, and a conservative progress save (advance `latestIsoDate` only
  after the full cursor walk completes). `processNymTx` maps each `EdgeTransactionRecord`
  into `StandardTx`.
- Native-unit string amounts are converted to major units via NYM's own
  `GET /api/partner/v1/currencies` decimals (keyed by currency code + tokenId), with a
  built-in fallback map for the current asset set if that fetch fails.
- Report timestamp keys off `createdDate` because NYM leaves `completedDate` null on most
  completed orders (observed 7 of 8 in the live sample).
- Status enum mapped `pending/processing/infoNeeded/expired/refunded/completed` to Edge
  StandardTx statuses.
- Registered `nymswap` in `src/queryEngine.ts` and `src/demo/partners.ts`. The `pluginId`
  is `nymswap`, matching the NYM swap plugin in edge-exchange-plugins.
- `test/nym.test.ts` - unit tests for `processNymTx` using real captured `EdgeTransactionRecord`
  payloads (amount decimal conversion, unknown-status degradation, null-txid tolerance,
  unknown-asset skip, non-numeric amount clamp).

Credentials are human/ops only for production (partner dashboard + production CouchDB
`reports_apps partnerIds.nymswap.apiKeys`); no production credential is fetched or set by
this code. The live verification used the GUI's existing swap key at the operator's request.

Open questions raised with the NYM team (see the Asana task):
- `completedDate` is null on completed orders - should reporting rely on `createdDate` for
  the settlement time, or will `completedDate` be populated?
- `sourceNetwork`/`destinationNetwork` casing is inconsistent (`ethereum` vs `BTC`/`NYM`) and
  does not map 1:1 to Edge plugin ids; chain plugin fields are left undefined for now.
- Confirm the production reporting key is the same key/scheme as the swap key, or a distinct
  reporting credential.

Testing:
- Live query against `GET /api/partner/v1/reports/transactions` under the GUI key returned
  21 real records (8 completed, 13 expired); the response matched the OpenAPI `TransactionReport`
  schema exactly and drove the cleaner design.
- Repo typecheck (`tsc` / `build.types`) clean of the new code; eslint clean on changed files.
- `test/nym.test.ts` passes (5/5).
- Note: the repo's existing mocha suite has 31 pre-existing failures (apiAnalytics + util
  date-math) unrelated to this change.

---

> [!NOTE]
> **Medium Risk**
> New external API integration affects reported swap amounts and timestamps (createdDate vs settlement); mis-scaled decimals are guarded by skipping unknown assets, but ops must configure the API key in CouchDB.
> 
> **Overview**
> Adds **NYM Swap (`nymswap`)** to the reports pipeline so Edge can ingest partner swap history from NYM’s Partner API.
> 
> A new plugin in `src/partners/nym.ts` pages `GET /api/partner/v1/reports/transactions` with `x-api-key`, a **5-day lookback**, cursor pagination, and retries; **`latestIsoDate` advances only after the full cursor walk finishes** so partial runs do not skip data. Native amount strings are scaled to major units using live **`/currencies`** decimals (with a static fallback), records map to **`StandardTx`** (status mapping, **`createdDate`** for timestamps because **`completedDate`** is often null), and bad rows are skipped instead of failing the whole page.
> 
> **`nymswap`** is wired into **`queryEngine`** and the demo partner list; **CHANGELOG** and **`test/nym.test.ts`** cover `processNymTx` conversion and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 834e37a42a9580586a8bb19c74e70dc227df93de. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/partners/nym.ts [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+  sourceCurrency: asString,
+  sourceAmount: asOptional(asString),
+  payinAddress: asOptional(asString),
+  payinTxid: asOptional(asString),
+
+  // Destination (payout) side.
+  destinationCurrency: asString,
+  destinationAmount: asOptional(asString),
+  payoutAddress: asOptional(asString),
+  payoutTxid: asOptional(asString)
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/partners/nym.ts]: Fixed in caf38d4. The six optional fields (sourceAmount, destinationAmount, payinAddress, payinTxid, payoutAddress, payoutTxid) now use asMaybe so a present-but-wrong-typed value degrades instead of throwing and aborting the lookback block. Amounts additionally accept a numeric encoding via a dedicated asNymAmount cleaner (degrades to 0). Added a unit test covering numeric amount + bad-typed field degradation.

File: src/partners/nym.ts [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+  other: 'other'
+}
+
+// Amounts may arrive as native-unit strings or plain numbers. Degrade an
+// unexpected encoding to 0 rather than throwing: a throw would escape
+// processNymTx into the page retry loop and abandon the whole lookback block.
+const asNymAmount = asMaybe((raw: unknown): number => {
+  if (typeof raw === 'number') return raw
+  return safeParseFloat(asString(raw))
+}, 0)
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/partners/nym.ts]: Valid — fixed in fbc8882. asNymAmount now clamps non-finite results to 0 via Number.isFinite, so non-numeric strings (safeParseFloat -> NaN) and raw NaN/Infinity never reach StandardTx. Unit test added.

File: src/partners/nym.ts [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+      log.error(String(e))
+      // Retry a few times with a delay to ride out throttling.
+      retry++
+      if (retry <= MAX_RETRIES) {
+        log.warn(`Snoozing ${5 * retry}s`)
+        await snooze(5000 * retry)
+      } else {
+        // Progress is saved oldest-to-newest, so it is safe to stop here.
+        break
+      }
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/partners/nym.ts:284]: Valid — fixed in fbc8882. Each order is processed in its own try/catch and a bad order is skipped (continue) instead of throwing out to the page retry, so already-pushed orders are never re-appended. Unit test added.

File: src/partners/nym.ts [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    } catch (e) {
+      log.error(String(e))
+      // Retry a few times with a delay to ride out throttling.
+      retry++
+      if (retry <= MAX_RETRIES) {
+        log.warn(`Snoozing ${5 * retry}s`)
+        await snooze(5000 * retry)
+      } else {
+        // Progress is saved oldest-to-newest, so it is safe to stop here.
+        break
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/partners/nym.ts:283]: Valid — fixed in fbc8882. Progress (latestIsoDate) now advances only when the query COMPLETES (empty page reached); an error-driven early exit keeps the prior saved value, so a partial fetch never skips unfetched orders regardless of API page ordering. Already-fetched orders are still returned and deduped by orderId.
